### ISSUE TITLE
pathprint

### ISSUE THREAD

**Issue by BioSok** [Bioconductor/Contributions#183]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/hidelab/pathprint

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Subversion][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/source-control/
[6]: http://bioconductor.org/developers/how-to/git-mirrors/

**Comment by mtmorgan** [comment]
@lshep will provide a review of this package; I notice that it does not import or depend on a single Bioconductor package. A key advantage to using Bioconductor is the opportunity for smooth interoperability across packages. I recommend that you provide and use an implementation that accepts Biobase ExpressionSet and GSEABase GeneSetCollection objects, and perhaps also the more modern SummarizedExperiment class. A discussion along these lines is associated with https://github.com/Bioconductor/Contributions/issues/170

**Comment by BioSok** [comment]
Dear @lshep , I do not recognize the build result "ABNORMAL", could you give me a direction as to what is missing?

**Comment by lshep** [comment]
This was an issue on our end having to do with the single package builder update.  I will manually kick off a rebuild of your package.

**Comment by lshep** [comment]
The "abnormal" is an error on our end from updating the single package builder. Once the update is complete I will manually kick off a build of your package

**Comment by vobencha** [comment]
Hi,
Please address the errors in the Single Package Builder output (maybe you're in the middle of doing this ...). Once the build report is clean I'll do a full review.
Valerie

**Comment by BioSok** [comment]
Dear @vobencha ,

There are no errors from the package build now :)

-Sokratis

**Comment by vobencha** [comment]
Hi,

I've done a first review. Please see suggestions and comments below.

(1) Add GEOquery to 'Suggests' in DESCRIPTION file.

(2) package size
The tar ball should occupy < 4MB on disk. See guidelines here
http://www.bioconductor.org/developers/package-guidelines/#correctness

~/software >du -h pathprint_0.99.23.tar.gz
25M    pathprint_0.99.23.tar.gz

There are too many, large data files in the package. Please consider using data from our experimental or annotation repos:
http://www.bioconductor.org/packages/release/BiocViews.html#___AnnotationData
http://www.bioconductor.org/packages/release/BiocViews.html#___ExperimentData

(3) UTF-8
R CMD build give this message:

** installing vignettes
   ‘exampleFingerprint.Rnw’ using ‘UTF-8’

If the vignette uses UTF-8 you need to specify that in the DESCRIPTION:
Encoding: UTF-8

(4) Add support for ExpressionSet objects as input
Your functions are just one step away from support ExpressionSet as input. This is a much more flexible way for the user to interact with your functions. For example, thresholdFingerprint() should accept an ExpressionSet as the 'SCE' argument and exprs2fingerprint() should accept an ExpressionSet as the 'exprs' argument.

Instead of this:
```
ALL.exprs<-exprs(ALL)
ALL.exprs.sub<-ALL.exprs[,1:5]
ALL.fingerprint<-exprs2fingerprint(exprs = ALL.exprs.sub,
    platform = "GPL8300",
    species = "human",
    progressBar = TRUE
)
```
a user should be able to do this:

```
ALL.fingerprint<-exprs2fingerprint(exprs = ALL,
    platform = "GPL8300",
    species = "human",
    progressBar = TRUE
)
```
The exprs2fingerprint() code simply needs to extract the exprs() data and pass it on down. If you want to run just a subset of data, the ExpressionSet object itself can be subset then passed to the function:

`sub <- ALL[1:4,]`

(5) vignette
i) add background
Please add a 'Background' section and describe the ideas behind Pathprint. It's not enough to reference a web site - the R package documentation needs to be complete on it's own. The background and justification [here](http://compbio.sph.harvard.edu/hidelab/pathprint/Pathprint.html
) needs to be in the vignette and / or man pages.

ii) change title
Please give the document a different title. Instead of
  \title{Pathway Fingerprinting - example script}
maybe 'Introduction to Pathway Fingerprinting' or 'Pathway Fingerprinting: a working example'.

Let me know if you have questions.

Valerie

**Comment by vobencha** [comment]
When you're ready for me to look at the package again please explain which changes have been made to address the issues raised in https://github.com/Bioconductor/Contributions/issues/183#issuecomment-261609343.

**Comment by BioSok** [comment]
Dear @vobencha ,
I have thought about your comments and responded as follows:
1)	Added GEOquery to 'Suggests' in DESCRIPTION file.
2)	Package size: Not much I can do here, unless the data files can be compressed further. The file that takes up the majority of the space is GEO.fingerprint.matrix.rda. This file contains a matrix representing a library of pathway fingerprints processed from a very large number of GEO datasets. This data is required for a key feature of the package; being able to search GEO samples for pathway profiles that closely match a new experiment. It is not possible to remove this file, nor can it be feasibly processed in a realistic timescale from any existing bioconductor package or repository.
3)	Added “Encoding: UTF-8” to the DESCRIPTION file.
4)	To allow for both expressionSet and matrix as the first input to the function, I adjusted the exprs2fingerprint function and updated its .Rd correspondingly.
5)	 i) Added Summary/Background/Methods/Pathway Sources section to the vignette
         ii) Adjusted the title

**Comment by vobencha** [comment]
Thanks for making the changes.

As for (2), the data can be split out into an experimental data package, examples here:
http://www.bioconductor.org/packages/release/data/experiment/

You can use the minfiData as a template package. Experimental data packages are fairly light - they don't need a vignette. The man pages should describe the data, how they were curated, the current R data type etc.

Once you have the data package ready, please make a separate github repo for it and upload it to this issue. The software package should 'Depend' on the data package; the data package should not have any dependencies on the software.

Valerie

**Comment by BioSok** [comment]
This is the separate repo for the data package [https://github.com/hidelab/pathprintGEOData.git](https://github.com/hidelab/pathprintGEOData.git) . Please inform me of any changes necessary.

**Comment by vobencha** [comment]
Instructions for adding a data package to a tracker issue are here:
https://github.com/Bioconductor/Contributions/blob/master/CONTRIBUTING.md#submitting-related-packages

**Comment by mtmorgan** [comment]
The correct way to add a package is with the following line

AdditionalPackage: https://github.com/hidelab/pathprintGEOData

Let's see if my comment triggers the addition & build. I'll then close the new issue that you opened. The reason for doing it this way is so that the same reviewer is assigned the data package.

**Comment by mtmorgan** [comment]
It looks like the comment that adds the package needs to come from @BioSok ; please comment with the AdditionalPackage line in my previous comment.

**Comment by BioSok** [comment]
AdditionalPackage: [https://github.com/hidelab/pathprintGEOData](https://github.com/hidelab/pathprintGEOData)

**Comment by mtmorgan** [comment]
The url you posted was in the form ` [https://github.com/hidelab/pathprintGEOData](https://github.com/hidelab/pathprintGEOData)` but should be just a plain-old-url, without markup

AdditionalPackage: https://github.com/hidelab/pathprintGEOData

**Comment by BioSok** [comment]
AdditionalPackage: https://github.com/hidelab/pathprintGEOData

**Comment by BioSok** [comment]
AdditionalPackage: https://github.com/hidelab/pathprintGEOData

**Comment by BioSok** [comment]
Hey,

Why am I getting the following errors in a data package?

* Checking vignette directory...
    This is a software package
    * ERROR: No 'vignettes' directory.
* Checking whether vignette is built with 'R CMD build'...
    * ERROR: Vignette must be built by 'R CMD build'. Please see the
      `Vignette Checks` section of the BiocCheck vignette.

**Comment by vobencha** [comment]
Hi,

Vignettes are not required for experimental data packages so you can ignore that error. I've reviewed the data package and taken a second look at the software.

## Data Package:
1 Please address these issues from R CMD check:

* checking for hidden files and directories ... NOTE
Found the following hidden files and directories:
  .git

* checking data for ASCII and uncompressed saves ... WARNING
 
  Note: significantly better compression could be obtained
        by using R CMD build --resave-data
                               old_size new_size compress
  GEO.fingerprint.matrix.RData   38.5Mb   19.9Mb       xz
  GEO.metadata.matrix.RData       3.1Mb    1.9Mb       xz

2) biocViews

These must come from the ExperimentData section, not software:

http://www.bioconductor.org/packages/release/BiocViews.html#___ExperimentData

3) DESCRIPTION

Please remove these lines:

_R_CHECK_FORCE_SUGGESTS_: FALSE
Packaged: 2016-12-12 14:01:00 UTC; SokratisKariotis

For a discussion on why we ask that _R_CHECK_FORCE_SUGGESTS_=TRUE see this thread:

https://stat.ethz.ch/pipermail/bioc-devel/2012-September/003638.html

4) object

Storing data separately from its metadata can be problematic. The dimensions (and documentation) can get out of sync when changes are made and it's difficult to correlate subsets of one object with the other. There are several Bioconductor objects that store data and metadata together to avoid these data management problems. One is the SummarizedExperiment class which has validity checks on the dimensions so the data don't drift. You can subset a single object with '[' which automatically subsets both data and metadata. It really is quite handy. Please take a look at the class in the SummarizedExperiment package and try putting the data in @assays and the metadata in @colData and @metadata, all accessable via getters / setters assays(), colData(), metadata(). You'll need SummarizedExperiment in Depends in DESCRIPTION.

First you need metadata for all samples. I think the y dimension of the data and the x dimension of the metadata should match, right?
```
> dim(GEO.fingerprint.matrix)
[1]    633 188390
> dim(GEO.metadata.matrix)
[1] 188502      7
```
Then something like,
```
library(SummarizedExperiment)
> colData <- GEO.metadata.matrix[-1]
> rownames(colData) <- GEO.metadata.matrix[,1]
> SummarizedExperiment(assays=list(fingerprint=GEO.fingerprint.matrix), colData=colData)
```
5) man page

Once the object and metadata are in a single object you can combine the man pages.

- Please describe how the data were curated, e.g., what is involved in pre-caclulating fingerprint vectors?

- Describe the stats of the data, e.g, ... "data are a 633 x 188390 integer matrix of pathways by samples ..."
```
> class(GEO.fingerprint.matrix[1])
[1] "integer"
> dim(GEO.fingerprint.matrix)
[1]    633 188390
> head(colnames((GEO.fingerprint.matrix)))
[1] "GSM100169" "GSM100197" "GSM100198" "GSM100200" "GSM100203" "GSM100212"
> head(rownames((GEO.fingerprint.matrix)))
[1] "Glycolysis / Gluconeogenesis (KEGG)"           
[2] "Citrate cycle (TCA cycle) (KEGG)"              
[3] "Pentose phosphate pathway (KEGG)"              
[4] "Pentose and glucuronate interconversions (KEGG)"
[5] "Fructose and mannose metabolism (KEGG)"        
[6] "Galactose metabolism (KEGG)"   
```
6) I'm seeing some inconsistencies between the data and metadata man pages (also mentioned in 4). The data are 633 x 188390 but the metadata man page mentions 158487 observations,
```
\format{
  A data frame with 158487 observations on the following 7 variables.
  \describe{
...
```
and yet the metadata object is
```
> dim(GEO.metadata.matrix)
[1] 188502      7
```
## Software package:

1) It looks like you haven't updated the software to use the new data package. Please remove data files in data/ and update man pages.

2) DESCRIPTION

- Remove these:
LazyLoad: yes
LazyData: yes
Packaged: 2012-06-14 15:08:12 UTC; GabrielAltschuler

- The data package is misspelled, should have capital 'D':
Depends: R (>= 3.4), pathprintGEOdata

Valerie

**Comment by drj11** [comment]
@vobencha you mention finding hidden file, in particular:

> checking for hidden files and directories ... NOTE
Found the following hidden files and directories:
.git

Surely that `.git` directory appears because you used `git clone` and didn't remove the `.git` directory that `git clone` creates?
It doesn't appear in the automated reports http://bioconductor.org/spb_reports/pathprintGEOData_buildreport_20170109112439.html

**Comment by vobencha** [comment]
Yes, thanks for pointing that out. I did use git clone and I did not remove the .git directory so please ignore that comment.
Valerie

**Comment by vobencha** [comment]
If you are ready for me to look at the packages again please let me know how comments in https://github.com/Bioconductor/Contributions/issues/183#issuecomment-271335410 have been addressed.
Valerie

**Comment by BioSok** [comment]
Dear @vobencha , here are the cahgnes I made per point made:
Data Package:
1. Used --resave-data to compress the two data files
2. biocViews updated properly
3. Mentioned lines removed
4. Data files were summarized to one (SummarizedGEO.Rdata) using SummarizedExperiment 
5. Single man page created with updated info
6. Incosistencies fixed by removing excessive metadata 
Software Package:
1. Removed files in data/ and man pages.
2. Depends: R (>= 3.4), pathprintGEOdata is correctly spelled now

**Comment by vobencha** [comment]
Hi,
Thanks for making the changes. A few more comments -

### data package
- add external_data_store.txt with the single line 'data' in it (same level as data/ and man/ directories in package structure). See source code for one of the experimental data packages here for an example:
http://www.bioconductor.org/packages/release/BiocViews.html#___ExperimentData
- remove the CheckVignettes: false line from DESCRIPTION
- man page
  -- add \alias{pathprintGEOData} to the man page so the user can find the data simply by knowing the name of the package
  -- change name of file from GEO.fingerprint_metadata.matrix.Rd to SummarizedGEO.Rd
  -- I don't see either GEO.fingerprint.matrix or GEO.metadata.matrix which are mentioned in the \usage section. Are these old?
- example section
This should demonstrate the use of the data package independent of the pathprint package. Show how to load and manipulate the data. Below I'm using load() since dispatch through data() isn't set up because you don't have the external_data_store.txt in place.
```
> load("SummarizedGEO.Rdata")  ## this should be data(SummarizedGEO)
```
Show what is loaded, some quick stats and maybe a few manipulations:
```
> objects()
[1] "result"
> class(result)
[1] "SummarizedExperiment"
attr(,"package")
[1] "SummarizedExperiment"
> dim(result)
[1]    633 188390
```
- vignette
add a brief vignette showing how to use the data with the pathprint package. Basically you could move the example from the man page to vignettes/ with a brief introduction.

## software package
- remove Date field from DESCRIPTION
- I'll check this again once the data package is functioning properly

**Comment by BioSok** [comment]
Dear @vobencha , it seems data() cannot find my dataset and the build is failing. Should I use load() instead?

**Comment by vobencha** [comment]
I'll have a look ...

**Comment by vobencha** [comment]
The dependency structure is fine - software Depends on data package and data package Suggests software. The SPB may not be capable of installing the packages in the correct order so don't worry about errors related the data package not being available.

When you've made the changes below please write a message clearly stating what you have done to address each point. Some of these are reoccurring problems; it is not a good use of time for me to review the pacakge(s) only to re-identify issues that have not been fixed.

## Data Package
1) Add import(SummarizedExperiment) to NAMESPACE
2) Change the file extension to .rda, e.g., data/pathprintSummarizedGEOData.rda
3) man page
    i) The package depends on SummarizedExperiment so you don't need to load in in the man page example. Please remove this:
```
# use the SummarizedExperiment library
library(SummarizedExperiment)
```
ii) The name and alias still have nothing to do with the name of the file or the objects that are loaded. Why are you referencing GEO.fingerprint_metadata.matrix?
```
\name{GEO.fingerprint_metadata.matrix}
\alias{GEO.fingerprint_metadata.matrix}
```
iii) Add an \alias for the name of the data file, the one you load in the \example section.
iv) Remove result output
The user can evaluate the man page be calling example(nameOfManPage) or by stepping through the code manually. It is not standard practice to include the expected output because it will then be shown twice. Please remove the output following all the code calls, e.g., remove [1] "result" from the call to objects():
```
# objects in environment
objects()
[1] "result"
```
4) vignette
    i) Rename directory with small 'v', "vignettes"
    ii) Remove library(SummarizedExperiment)
    iii) Please format code to max width <= 80 characters

## Software Package

1) R CMD build
It is your responsibility to run both R CMD build and check on the data package and the software pacakge. When I fixed the file extension on the data package, installed it and then ran R CMD build on the software I get this -
```
~/proj/git/software >R-dev CMD build pathprint/
* checking for file ‘pathprint/DESCRIPTION’ ... OK
* preparing ‘pathprint’:
* checking DESCRIPTION meta-information ... OK
* installing the package to build vignettes
* creating vignettes ... ERROR
...
Error: processing vignette 'exampleFingerprint.Rnw' failed with diagnostics:
 chunk 3 (label = fingerprint)
Error in match(x, table, nomatch = 0L) : object 'chipframe' not found
Execution halted
```

Valerie

**Comment by BioSok** [comment]
Dear @vobencha , I am sorry for the repetition it seems I pushed a previous version where the mentioned issues were not addressed yet. To sum up here is what I did:

Data package:
1.	`import(SummarizedExperiment)` added to NAMESPACE 

2.	the data file is now named pathprintSummarizedGEOData.rda

3.	i) Removed:  
`# use the SummarizedExperiment`
`library(SummarizedExperiment)`

ii/iii) I now only have:  `\alias{pathprintSummarizedGEOData}` in man file

iv) all result outputs removed

4.	i) Directory named: vignettes
ii) removed `library(SummarizedExperiment)` from usingPathprintGEOData.Rnw
iii) formatted the code lines that were close to 80 chars   

Software package:

I made the required changes in the package and now the software package, as well as the above data package, are building and checking without errors on my machine. However, when I pushed the last version of the software package I received the following error only on the tokay2 host: 

> GEO.fingerprint.matrix = assays(result)$fingerprint
Error: cannot allocate vector of size 915.9 Mb
Execution halted

**Comment by vobencha** [comment]
Hi,
Thank you for outlining the changes that were made. Still a few more things to address.

1) windows error
The software packages are checked on both 32 and 64 bit architectures on windows. The error you're seeing is the 32 bit run. We need to do 2 things, the first is to save the data with a more efficient compression and the second is to use a smaller subset of the object in the example if possible, e.g., 
assays(result)$fingerprint[1:1000]

Please resave the data object with save(..., compress='xz'). It looks like we can get a reduction of 50% with this compression:

> load ("pathprintSummarizedGEOData.rda")
> save(result, file="compressed_result.rda", compress='xz')
> q()
~/proj/git/software/pathprintGEOData/data >du -h *
22M	compressed_result.rda
42M	pathprintSummarizedGEOData.rda


2) R CMD build / check output
Please address all issues below. Let me know if you have questions.

* checking dependencies in R code ... NOTE
Package in Depends field not imported from: 'pathprintGEOData'
  These packages need to be imported from (in the NAMESPACE file)
  for when this namespace is loaded but not attached.

* Undefined global functions or variables:
  aggregate chipframe genesets median na.omit p.adjust phyper
  platform.thresholds pnorm sd setTxtProgressBar txtProgressBar
Consider adding
  importFrom("stats", "aggregate", "median", "na.omit", "p.adjust",
             "phyper", "pnorm", "sd")
  importFrom("utils", "setTxtProgressBar", "txtProgressBar")
to your NAMESPACE file.

* checking Rd files ... NOTE
prepare_Rd: hyperPathway.Rd:26-28: Dropping empty section \details
prepare_Rd: hyperPathway.Rd:29-35: Dropping empty section \value
prepare_Rd: hyperPathway.Rd:43-45: Dropping empty section \note
prepare_Rd: hyperPathway.Rd:46-48: Dropping empty section \seealso
prepare_Rd: hyperPathway.Rd:49-51: Dropping empty section \examples

* checking for missing documentation entries ... WARNING
Undocumented data sets:
  'GEO.fingerprint.matrix' 'GEO.metadata.matrix'

* Checking exported objects have runnable examples...
    * NOTE: Consider adding runnable examples to the following man
      pages which document exported objects:
      hyperPathway.Rd

**Comment by BioSok** [comment]
Dear @vobencha , 

1. I have compressed the files as suggested and the automated reply did not contain the previous size warning any more.

2. i) Added imports (pathprintGEOData, stats and utils) in the NAMESPACE file.
   ii) Removed empty sections from Rd files
  iii) Added aliases for 'GEO.fingerprint.matrix' and 'GEO.metadata.matrix'
  iv) hyperPathway.R and its .Rd file were removed as they were not needed by the package. If decided otherwise I will provide a new man page with a working example

My last pathprint package commit yielded only the warning (only in the automated bioconductor check)

* Checking R Version dependency...
    * WARNING: Update R version dependency from 3.3 to 3.4.

However, changing the dependency causes an error in my machine.

**Comment by BioSok** [comment]
Dear @vobencha, I I changed the R version dependency to 3.4 (as prompted by the warning message) and the package got an "OK" status. What are the next steps?

**Comment by vobencha** [comment]
I'll take a look at this tomorrow.
Valerie

**Comment by vobencha** [comment]
When I run R CMD check on the data package I see 3 warnings and 2 notes. Everything needs to be fixed except this note:
* checking installed package size ... NOTE
  installed size is 80.4Mb
  sub-directories of 1Mb or more:
    data  80.2Mb

Valerie

**Comment by BioSok** [comment]
Dear @vobencha , 
I address all warnings and currently both packages have the OK label.
-Sokratis

**Comment by vobencha** [comment]
Thank you. I'm marking the software and data packages as approved. You will get an email in the next week with svn credentials and next steps.

Valerie
